// @cof-org/mcp/dist/adapters/mock/mock-config.js
/**
 * Mock Configuration
 * Configuration for mock adapter behavior
 */
export class MockConfig {
    options;
    constructor(options = {}) {
        // Disable error simulation in test environment
        const isTestEnvironment = process.env.NODE_ENV === 'test' ||
            process.env.NODE_ENV === 'development';
        this.options = {
            minLatency: 50,
            maxLatency: 200,
            fixedLatency: undefined,
            errorRate: isTestEnvironment ? 0 : 0.01, // Disable errors in test/dev
            operationErrors: {},
            persistData: true,
            dataSize: 1000,
            realTimeUpdates: false,
            randomizeData: true,
            ...options
        };
    }
    /**
     * Get simulated latency in milliseconds
     */
    getLatency() {
        if (this.options.fixedLatency !== undefined) {
            return this.options.fixedLatency;
        }
        const min = this.options.minLatency || 0;
        const max = this.options.maxLatency || 0;
        return Math.floor(Math.random() * (max - min + 1)) + min;
    }
    /**
     * Check if operation should simulate an error
     */
    shouldSimulateError(operation) {
        // Check operation-specific error rate first
        const operationRate = this.options.operationErrors?.[operation];
        if (operationRate !== undefined) {
            return Math.random() < operationRate;
        }
        // Fall back to global error rate
        return Math.random() < (this.options.errorRate || 0);
    }
    /**
     * Get configuration summary for health checks
     */
    getSummary() {
        return {
            latency: this.options.fixedLatency !== undefined
                ? `fixed ${this.options.fixedLatency}ms`
                : `${this.options.minLatency}-${this.options.maxLatency}ms`,
            errorRate: `${(this.options.errorRate || 0) * 100}%`,
            dataSize: this.options.dataSize,
            persistData: this.options.persistData,
            realTimeUpdates: this.options.realTimeUpdates
        };
    }
    /**
     * Get data size configuration
     */
    getDataSize() {
        return this.options.dataSize || 1000;
    }
    /**
     * Check if data should be persisted
     */
    shouldPersistData() {
        return this.options.persistData !== false;
    }
    /**
     * Check if data should be randomized
     */
    shouldRandomizeData() {
        return this.options.randomizeData !== false;
    }
    /**
     * Check if real-time updates are enabled
     */
    hasRealTimeUpdates() {
        return this.options.realTimeUpdates === true;
    }
    /**
     * Set error rate for specific operation
     */
    setOperationErrorRate(operation, rate) {
        if (!this.options.operationErrors) {
            this.options.operationErrors = {};
        }
        this.options.operationErrors[operation] = rate;
    }
    /**
     * Update configuration options
     */
    updateOptions(newOptions) {
        this.options = { ...this.options, ...newOptions };
    }
}
