// @cof-org/mcp/dist/utils/logger.js
/**
 * Logger utility - Enhanced with structured logging
 * Logs to stderr to avoid polluting stdio transport
 * Maintains backward compatibility while adding structured logging
 */
import { getLogger } from '../logging/structured-logger.js';
export class Logger {
    static level = 'info';
    static levels = {
        debug: 0,
        info: 1,
        warn: 2,
        error: 3,
    };
    static structuredLogger = null;
    static useStructured = false;
    static init(level, useStructured = false, logConfig) {
        this.level = level;
        this.useStructured = useStructured;
        if (useStructured) {
            this.structuredLogger = getLogger('cof-mcp', logConfig);
        }
    }
    static shouldLog(level) {
        return this.levels[level] >= this.levels[this.level];
    }
    static format(level, message, ...args) {
        const timestamp = new Date().toISOString();
        const formatted = `[${timestamp}] [${level.toUpperCase()}] ${message}`;
        if (args.length > 0) {
            const additional = args.map(arg => {
                if (arg instanceof Error) {
                    return `${arg.message}\n${arg.stack}`;
                }
                if (typeof arg === 'object') {
                    try {
                        return JSON.stringify(arg, null, 2);
                    }
                    catch {
                        return String(arg);
                    }
                }
                return String(arg);
            }).join(' ');
            return `${formatted} ${additional}`;
        }
        return formatted;
    }
    static debug(message, ...args) {
        if (this.shouldLog('debug')) {
            if (this.useStructured && this.structuredLogger) {
                const context = args.length > 0 ? { args } : undefined;
                this.structuredLogger.debug(message, context);
            }
            else {
                console.error(this.format('debug', message, ...args));
            }
        }
    }
    static info(message, ...args) {
        if (this.shouldLog('info')) {
            if (this.useStructured && this.structuredLogger) {
                const context = args.length > 0 ? { args } : undefined;
                this.structuredLogger.info(message, context);
            }
            else {
                console.error(this.format('info', message, ...args));
            }
        }
    }
    static warn(message, ...args) {
        if (this.shouldLog('warn')) {
            if (this.useStructured && this.structuredLogger) {
                const context = args.length > 0 ? { args } : undefined;
                this.structuredLogger.warn(message, context);
            }
            else {
                console.error(this.format('warn', message, ...args));
            }
        }
    }
    static error(message, ...args) {
        if (this.shouldLog('error')) {
            if (this.useStructured && this.structuredLogger) {
                // Handle Error objects specially
                if (args.length === 1 && args[0] instanceof Error) {
                    this.structuredLogger.error(message, args[0]);
                }
                else {
                    const context = args.length > 0 ? { args } : undefined;
                    this.structuredLogger.error(message, context);
                }
            }
            else {
                console.error(this.format('error', message, ...args));
            }
        }
    }
    // New structured logging methods
    static getStructuredLogger() {
        return this.structuredLogger;
    }
    static enableStructuredLogging(logConfig) {
        this.useStructured = true;
        if (!this.structuredLogger) {
            this.structuredLogger = getLogger('cof-mcp', logConfig);
        }
    }
    static disableStructuredLogging() {
        this.useStructured = false;
    }
}
