// @cof-org/mcp/dist/utils/timeout.js
/**
 * Timeout utilities for enforcing performance configuration
 */
export class TimeoutHandler {
    static config = null;
    /**
     * Set the timeout configuration from server config
     */
    static setConfig(timeoutConfig) {
        this.config = timeoutConfig;
    }
    /**
     * Wrap an operation with a timeout based on configuration
     * @param operation - The async operation to execute
     * @param timeoutType - Type of timeout ('request' or 'adapter')
     * @param customTimeout - Override timeout value
     * @returns Promise resolving to the operation result
     */
    static async withTimeout(operation, timeoutType, customTimeout) {
        const timeout = customTimeout || (this.config ? this.config[timeoutType] : timeoutType === 'request' ? 30000 : 5000);
        return new Promise((resolve, reject) => {
            const timeoutId = setTimeout(() => {
                reject(new Error(`Operation timed out after ${timeout}ms`));
            }, timeout);
            operation()
                .then((result) => {
                clearTimeout(timeoutId);
                resolve(result);
            })
                .catch((error) => {
                clearTimeout(timeoutId);
                reject(error);
            });
        });
    }
    /**
     * Create a timeout promise that rejects after the specified time
     * @param ms - Milliseconds to wait before rejecting
     * @param message - Error message to use when timing out
     */
    static createTimeout(ms, message = 'Operation timed out') {
        return new Promise((_, reject) => {
            setTimeout(() => reject(new Error(message)), ms);
        });
    }
    /**
     * Get configured timeout for a specific type
     * @param timeoutType - Type of timeout to get
     * @returns Timeout in milliseconds
     */
    static getTimeout(timeoutType) {
        if (!this.config) {
            return timeoutType === 'request' ? 30000 : 5000;
        }
        return this.config[timeoutType];
    }
}
